### ISSUE TITLE
Introduce & enforce Versioning for APIs

### ISSUE THREAD

**Issue by Malyuk-A** [oakestra/oakestra#333]
## Short
The Oakestra APIs should have strong versioning to avoid unexpected breaking changes and easier extendability/modifiability.

## Proposal
Rather self-explanatory. We need to add/make sure to have comprehensive & documented rules for using versioning and enforce their usage in our codebases.

Context: @giobart and I discussed the current state of our codebase and he mentioned that this is a major thing we should work on.
@melkodary @TheDarkPyotr 

## Impact
Every component that exposes APIs, Documentation

## Development time
Depends on how we split the effort. Including documentation, initial discussions for best practices to follow, etc. This can take a couple of weeks. The implementation itself is rather quick. (Figure out once, apply everywhere)

## Status
Looking for discussion/feedback, best practices, and how to split the workload.

## Checklist
 
- [ ] Discussed
- [ ] Documented
- [ ] Implemented
- [ ] Tested

**Comment by Malyuk-A** [comment]
That is a fantastic comment @TheDarkPyotr !
To be fair I think Giovanni & I were only talking about API versioning. Your comment covers a suite of necessary steps that include and go beyond versioning. I really appreciate your effort!

I totally agree with your assessment.
I would use your comment as a list to generate further tickets instead of having everything in one place.

To add some more ideas to your list:
- Python typing/type-hints would be massively helpful IMO for developers to figure out what is going on. (It is not great to read code where you do not know what exactly the input params are and what gets returned)
- Regarding 2.i) I highly recommend using [this](https://docs.python.org/3/library/http.html#http-status-codes) instead of passing around the HTTP response codes as string or ints. Thus combined with type hints one can instantly know what this "status" object is. (I am currently looking into service statuses and oftentimes I do not know if this status object is an HTTP status, a service status, or some other status).

- There are a lot of ideas and possible improvements here. The trick is not to get overwhelmed and do small steps of improvement one at a time. (Especially because most of the Oakestra devs are thesis students who can only put as much time into general codebase improvements as their circumstances allow.)

**Comment by TheDarkPyotr** [comment]
Thank you @Malyuk-A! 🙂
Totally agree that the proposal go a bit too much beyond the API versioning. I think that establish a broader _"approach"_ (_that hardly becomes effectively implemented in its entirely from the conception_) can clarify the path to follow for new future components _(on the long run, obtaining a small-scale document similar to [this](https://opensource.zalando.com/restful-api-guidelines/))_.

Yours are really great suggestions! 🔥

Absolutely agree that time and resources are limited so it's important to make these improvements manageable.
Focusing on action-oriented points,  starting simple maybe we can consider using:
 - As you suggested, Python typing would be a fantastic improvement in term of clarity: using it in combination with [Pydantic](https://github.com/pydantic/pydantic) should be a very good start (_mostly towards points 2 and 3_) to have a consistent object representation
- [Flask-Rebar](https://flask-rebar.readthedocs.io/en/latest/quickstart/api_versioning.html) for minimal versioning of API

**Comment by giobart** [comment]
@Mjaethers check how much effort is needed when integrating this with the gateway. Should this be done before or after?

**Comment by Mjaethers** [comment]
With respect to the gateway this is not a huge change, but after makes sense, as some of the work is already done.

I've already had to edit many hard-coded API calls so that they can be flexibly route through the respective kong gateway, only when necessary. This affects the dashboard and the inter component calls. (CLI is still outstanding)

After the gateway changes have been made, implementing a versioning system to the API endpoints also means changing the routing tables for kong. This is a mechanical task that can be automated fairly easily.
The kong routing tables can be configured to accept any API version using regex, meaning we don't have to update them every time we change the API version.

The addons system and resource abstractor endpoints already have a versioning system (although it has never been used).

**Comment by giobart** [comment]
Blocked by #526
